ocaml/xapi/xapi_vm.ml

```diff
@@ -236,14 +236,54 @@ let start_on  ~__context ~vm ~host ~start_paused ~force =
 	assert_host_is_localhost ~__context ~host;
 	start ~__context ~vm ~start_paused ~force
 
-let hard_reboot ~__context ~vm =
-	update_vm_virtual_hardware_platform_version ~__context ~vm;
-	Xapi_xenops.reboot ~__context ~self:vm None
+
+(* Nb, we're not using the snapshots returned in 'Event.from' here because
+   the tasks might get deleted. The standard mechanism for dealing with
+   deleted events assumes you have a full database replica locally, and
+   deletions are handled by checking your valid_ref_counts table against
+   your local database. In this case, we're only interested in a subset of
+   events, so this mechanism doesn't work. There will only be a few outstanding
+   tasks anyway, so we're safe to just iterate through the references when an
+   event happens - ie, we use the event API simply to wake us up when something
+   interesting has happened. *)
+let wait_for_tasks ~__context ~tasks =
+  let our_task = Context.get_task_id __context in
+  let classes = List.map (fun x -> Printf.sprintf "task/%s" (Ref.string_of x)) (our_task::tasks) in
+
+  let rec process token =
+    TaskHelper.exn_if_cancelling ~__context; (* First check if _we_ have been cancelled *)
+    let statuses = List.filter_map (fun task -> try Some (Db.Task.get_status ~__context ~self:task) with _ -> None) tasks in
+    let unfinished = List.exists (fun state -> state = `pending) statuses in
+    if unfinished
+    then begin
+      let from = Helpers.call_api_functions ~__context
+          (fun rpc session_id -> Client.Event.from ~rpc ~session_id ~classes ~token ~timeout:30.0) in
+      debug "Using events to wait for tasks: %s" (String.concat "," classes);
+      let from = Event_types.event_from_of_rpc from in
+      process from.Event_types.token
+    end else
+      ()
+  in
+  process ""
+
+let cancel ~__context ~vm ~ops =
+  let cancelled = List.filter_map (fun (task,op) ->
+    if List.mem op ops then begin
+      debug "Cancelling VM.%s for VM.hard_shutdown/reboot" (Record_util.vm_operation_to_string op);
+      Helpers.call_api_functions ~__context
+        (fun rpc session_id -> try Client.Task.cancel ~rpc ~session_id ~task:(Ref.of_string task) with _ -> ());
+      Some (Ref.of_string task)
+    end else None
+  ) (Db.VM.get_current_operations ~__context ~self:vm) in
+  wait_for_tasks ~__context ~tasks:cancelled
+
 
 let hard_shutdown ~__context ~vm =
+	cancel ~__context ~vm ~ops:[ `clean_shutdown; `clean_reboot; `hard_reboot; `pool_migrate; `call_plugin; `suspend ];
```

Comment:
Do we need to worry about current operations on devices of the VM (VIFs, VBDs, ...)?